src/sync.jl

```diff
@@ -0,0 +1,131 @@
+module Sync
+
+    @enum SyncMode mutex pin disabled
+
+    const gap_lock = ReentrantLock()
+    const pinned_thread = Ref{Int}(1)
+
+    @inline function lock()
+        Base.lock(gap_lock)
+    end
+
+    @inline function unlock()
+        Base.unlock(gap_lock)
+    end
+
+    @inline function check_lock()
+        @assert gap_lock.locked_by === Base.current_task()
+    end
+
+    @inline function check_pinned()
+        @assert Threads.threadid() == pinned_thread[]
+    end
+
+    # To switch between multi-threaded and single-threaded mode, we
+    # define functions that install the appropriate handlers for sync()
+    # etc.
+    #
+    # This is necessary because otherwise precompilation would fix
+    # the mode at whatever state it was during precompilation. Thus,
+    # initially loading GAP.jl in single-threaded mode would also keep
+    # synchronization off in multi-threaded mode.
+    #
+    # However, Julia tracks function dependencies. If a function changes
+    # upon which another depends, both are being recompiled. Thus,
+    # by installing the proper version during __init__(), we force
+    # selective recompilation of the affected functions as needed.
+
+    function _mode_mutex()
```

Comment:
This is unused in practice, it seems (as nothing calls `switch()`, it's single caller). It should either be removed or else a brief source comment be added which explains why it's here.